### PR TITLE
Adapt Dockerfile to using virtual framebuffers for notebook visualization

### PR DESCRIPTION
**Pull request by jorgensd**
To be able to use `pyvista` and other VTK based visualization modules in notebooks, we need a virtual framebuffer to visualize in.
Minimal test dockerfile has been built in https://github.com/jorgensd/PVGeo-Examples/blob/master/docker/Dockerfile
and can be found at: `dokken92/dolfinx_custom:test_lab_env`

### CODE REVIEW

**Review by jhale — CHANGES_REQUESTED** [review]
Looks good.

If the entrypoint has become too unwieldy to have in the dockerfile directly (preferred but I can see the issue here) then it would be best just to combine it all into one script e.g. 'launch_lab.sh' rather than combine the entrypoint and script approaches.

**Comment by garth-wells** [comment]
`ipyvtk_simple` (or more likely its dependency `ipycanvas`) seems to break things with JupyterLab 3. If `ipycanvas` is enabled it breaks all widgets, e.g. `itkwidgets` for me.

**Comment by jorgensd** [comment]
> `ipyvtk_simple` (or more likely its dependency `ipycanvas`) seems to break things with JupyterLab 3. If `ipycanvas` is enabled it breaks all widgets, e.g. `itkwidgets` for me.

Could you make your suggested changes to this PR, as I have not been able to get `itkwidgets` to work with any configuration on my system.

**Comment by garth-wells** [comment]
@jorgensd See https://github.com/FEniCS/dolfinx/tree/garth/itkwidgets-lab

**Comment by jorgensd** [comment]
> @jorgensd See https://github.com/FEniCS/dolfinx/tree/garth/itkwidgets-lab

On linux (ubuntu 20.04) following your instructions, I cannot use ITK.
Following Dockerfile 
```
from dolfinx/lab

RUN wget -qO - https://deb.nodesource.com/setup_15.x | bash && \
    apt-get -qq update && \
    apt-get install -y libgl1-mesa-dev xvfb nodejs && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*

RUN pip3 install --no-cache-dir jupyter jupyterlab pyvista ipygany itkwidgets
RUN jupyter labextension install @jupyter-widgets/jupyterlab-manager jupyter-matplotlib jupyterlab-datawidgets itkwidgets ipygany

EXPOSE 8888/tcp
ENV SHELL /bin/bash

ENTRYPOINT ["jupyter", "lab", "--ip", "0.0.0.0", "--no-browser", "--allow-root"]
```
with run command:
```
 docker run -v $PWD:/root/shared --init -p 8888:8888  name_of_image
```
causes `Error displaying widget: model not found` 
```
plotter = pyvista.PlotterITK()
plotter.add_mesh(grid)
plotter.show(True)
```
which I think is due to: https://github.com/InsightSoftwareConsortium/itkwidgets/issues/397

File: docker/launch_lab.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+set -x
+export DISPLAY=:99.0
+Xvfb :99 -screen 0 1024x768x24 > /dev/null 2>&1 &
+set +x
+jupyter lab --ip 0.0.0.0 --no-browser --allow-root
+exec "$@"
```
Comment (jhale) [inline docker/launch_lab.sh:7]: This line isn't necessary anymore.

**Comment by garth-wells** [comment]
```
FROM dolfinx/dolfinx
LABEL description="DOLFIN-X Jupyter Lab"

WORKDIR /root

# Dependencies for pyvista and related packages
RUN wget -qO - https://deb.nodesource.com/setup_15.x | bash && \
    apt-get -qq update && \
    apt-get install -y libgl1-mesa-dev xvfb nodejs && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*

RUN pip3 install --no-cache-dir jupyter jupyterlab pyvista ipygany itkwidgets
RUN jupyter labextension install @jupyter-widgets/jupyterlab-manager jupyter-matplotlib jupyterlab-datawidgets itkwidgets ipygany

EXPOSE 8888/tcp
ENV SHELL /bin/bash

ENTRYPOINT ["jupyter", "lab", "--ip", "0.0.0.0", "--no-browser", "--allow-root"]
```
and
```
import pyvista as pv
mesh = pv.Sphere()
z = mesh.points[:, 2]
pl = pv.PlotterITK()
pl.add_mesh(mesh, scalars=z, smooth_shading=True)
pl.show(True)
```
works.

**Comment by jorgensd** [comment]
> ```
> FROM dolfinx/dolfinx
> LABEL description="DOLFIN-X Jupyter Lab"
> 
> WORKDIR /root
> 
> # Dependencies for pyvista and related packages
> RUN wget -qO - https://deb.nodesource.com/setup_15.x | bash && \
>     apt-get -qq update && \
>     apt-get install -y libgl1-mesa-dev xvfb nodejs && \
>     apt-get clean && \
>     rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
> 
> RUN pip3 install --no-cache-dir jupyter jupyterlab pyvista ipygany itkwidgets
> RUN jupyter labextension install @jupyter-widgets/jupyterlab-manager jupyter-matplotlib jupyterlab-datawidgets itkwidgets ipygany
> 
> EXPOSE 8888/tcp
> ENV SHELL /bin/bash
> 
> ENTRYPOINT ["jupyter", "lab", "--ip", "0.0.0.0", "--no-browser", "--allow-root"]
> ```
> 
> and
> 
> ```
> import pyvista as pv
> mesh = pv.Sphere()
> z = mesh.points[:, 2]
> pl = pv.PlotterITK()
> pl.add_mesh(mesh, scalars=z, smooth_shading=True)
> pl.show(True)
> ```
> 
> works for me.

I guess you are on a mac? Using your Dockerfile still gives me the errormessage above.
error_widget

**Comment by mscroggs** [comment]
Me and @jorgensd just tried it on my computer and it worked ok

**Comment by jhale** [comment]
Built the above `Dockerfile` on my computer (Mac OS, latest Chrome) and it's broken.

@mscroggs Could you share your image as a .tar.gz?

**Comment by jorgensd** [comment]
> Built the above `Dockerfile` on my computer (Mac OS, latest Chrome) and it's broken with the same error message `Error displaying widget: model not found`.
> 
> @mscroggs Could you share your image as a .tar.gz?

We have decided to remove ITKWidgets for now, as it adds on 3GB to the size of the Dockerfile. This can be installed by users in custom docker-images if they desire to use it.

It also seems like jupyterlab 3.x is breaking ITKWidgets on a subset of computers. Using normal jupyter notebook resolves these issues for me. Hopefully all the different visualization packages will get full support of jupyterlab eventually.

**Review by garth-wells — APPROVED** [review]

**Pull request merged**